### ISSUE TITLE
Check if organelle shader blendmode needs changes

### ISSUE THREAD

**Issue by hhyyrylainen** [Revolutionary-Games/Thrive#1070]
It seems that the blending is not very good ATM and when the organelle draw order changes around there is visible difference when the organelles swap the order they are drawn in.

**Comment by Kasterisk** [comment]
I also got this same problem when I set the iron chunks to have a custom shader material and I _think_ it got fixed by adding `render_mode depth_draw_alpha_prepass;` on the shader.

**Comment by hhyyrylainen** [comment]
That does seem to make the organelle shadows etc. appear and look good. But it makes the organelle render order even more noticeable. So maybe some priority etc. stuff is needed.

![2021-07-16_16 32 29 8651](https://user-images.githubusercontent.com/3796411/125956311-f850c51f-c0f9-49fc-8c3f-88109592f734.png)
![2021-07-16_16 32 46 6590](https://user-images.githubusercontent.com/3796411/125956318-f49429e5-e05e-4433-9619-e8a27962a397.png)
![2021-07-16_16 32 50 2666](https://user-images.githubusercontent.com/3796411/125956331-31e2f0c0-bd96-42c3-8fda-cd3457716afd.png)


Help from a graphics programmer here would be awesome.

**Comment by Nunez2196** [comment]
so I read the Godot docs and Godot doesn't have order independent transparency which means this sorting issue we're experiencing can't really be fixed unless you make separate materials for separate instances (not just different types), and set different render priorities for various organelle instances that may come in contact with each other. This is because transparent objects are drawn different than opaque ones. This does mean however, we could fix at least some of the problem by making as many organelles opaque as possible, or we could write code to make sure organelles never touch each other. Currently almost every single one has transparency somewhere and the code squishes organelles together at some point I think.

source: https://docs.godotengine.org/en/stable/tutorials/3d/3d_rendering_limitations.html

**Comment by Nunez2196** [comment]
blendmode is for how pixel color value is mixed. There's mix, multiply, add, subtract. Not sure how that relates to the issue described or what it is you want to do with that to be honest with you.

**Comment by Nunez2196** [comment]
render_mode depth_draw_alpha_prepass; fixes problems with the transparency sorting being silly with the meshes themselves (the Rusticyanin was especially egregious, just spin the camera around it without that render mode) but I don't think that will fix the bigger issue of draw order.

**Comment by Nunez2196** [comment]
One issue I found in the shader themselves is that somebody made it where some the vertices of a given organelle in world space can change independent of the Node's position to make them wiggle. This causes issues too since the sorting goes off of Node not the position of actual vertices like with opaque objects.

**Comment by hhyyrylainen** [comment]
> One issue I found in the shader themselves is that somebody made it where some the vertices of a given organelle in world space can change independent of the Node's position to make them wiggle. This causes issues too since the sorting goes off of Node not the position of actual vertices like with opaque objects.

It was like that back when I joined the project. I would hazard a guess that it was considered more performant to apply the animation in the shader, rather than have the CPU side code calculate the organelle wiggle and update the model positions.

> so I read the Godot docs and Godot doesn't have order independent transparency which means this sorting issue we're experiencing can't really be fixed unless you make separate materials for separate instances (not just different types), and set different render priorities for various organelle instances that may come in contact with each other.

Which is a shame, because I think I've seen in some rendering engine a feature where you can add a distance bias, depth bias, or something like that to transparent objects, which makes the rendering engine *think* their distance to the camera is different than it really is. This works like the render priority *but* is also affected by the real distance to the camera, so I think with a feature like this we could avoid visual artifacts by applying slight random distance modification to all the different organelles to get them to render in a consistent order while a cell spins around.

**Comment by Nunez2196** [comment]
> Which is a shame, because I think I've seen in some rendering engine a feature where you can add a distance bias, depth bias, or something like that to transparent objects, which makes the rendering engine _think_ their distance to the camera is different than it really is. This works like the render priority _but_ is also affected by the real distance to the camera, so I think with a feature like this we could avoid visual artifacts by applying slight random distance modification to all the different organelles to get them to render in a consistent order while a cell spins around.

Ok, so basically you would like the organelles to have very slightly different distances to the camera so the sorting works normally? It might work as long as the distances don't get too severe or something to where players can tell there's organelles pretty far down below the membrane.

**Comment by hhyyrylainen** [comment]
> Ok, so basically you would like the organelles to have very slightly different distances to the camera so the sorting works normally? It might work as long as the distances don't get too severe or something to where players can tell there's organelles pretty far down below the membrane.

Doing it with actual distances doesn't work due to it affecting the rendering. I'm talking about *pretend* distances that are only used for depth sorting. This seems to be a feature missing from Godot.

**Comment by Nunez2196** [comment]
> This seems to be a feature missing from Godot.

That is correct, Godot doesn't support that. So the only solution is probably what I stated earlier about using various materials with different priorities. Unless you can think of something else or want to make things opaque.
